crates/jstz_core/src/kv/outbox.rs

```diff
@@ -0,0 +1,461 @@
+        }),))(input)
+    }
+}
+
+impl From<OutboxMessage> for OutboxMessageFull<OutboxMessage> {
+    fn from(message: OutboxMessage) -> Self {
+        match message {
+            OutboxMessage::Withdrawal(_) => {
+                OutboxMessageFull::AtomicTransactionBatch(message)
+            }
+        }
+    }
+}
+
+/// Represents a pending outbox queue stored as part of the
+/// trasaction's snapshot.
+#[derive(Debug, Default)]
+pub struct OutboxQueueSnapshot(Vec<OutboxMessage>);
+
+impl OutboxQueueSnapshot {
+    pub fn extend(&mut self, queue: OutboxQueueSnapshot) {
+        self.0.extend(queue.0)
+    }
+
+    pub fn queue_message(&mut self, message: OutboxMessage) {
+        self.0.push(message)
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct OutboxQueueMeta {
+    /// Current queue length between the rollup and snapshots'
+    /// outbox queues
+    queue_len: u32,
+    /// Maximum capacity of the rollup outbox queue
+    max: u32,
+}
+
+impl OutboxQueueMeta {
+    // FIX: Unfortunately, the RollupOutboxQueue does not expose any methods to
+    // read its metadata so use this workaround for now.
+    pub fn read_meta(rt: &impl Runtime) -> Result<Option<OutboxQueueMeta>> {
+        Storage::get::<OutboxQueueMeta>(rt, &OUTBOX_QUEUE_ROOT)
+    }
+
+    fn write_meta(&self, rt: &mut impl Runtime) -> Result<()> {
+        Storage::insert(rt, &OUTBOX_QUEUE_ROOT, self)
+    }
+}
+
+/// An outbox queue that composes outbox queue snapshots
+/// and `ROLLUP_OUTBOX_QUEUE` while maintaining the invariants
+/// of the kernel outbox.
+#[derive(Debug)]
+pub struct OutboxQueue {
+    /// Metadata for the outbox queue
+    meta: OutboxQueueMeta,
+
+    /// Rollup outbox queue
+    rollup_outbox_queue: RollupOutboxQueue,
+}
+
+impl OutboxQueue {
+    fn write_outbox_message(
+        rt: &mut impl Runtime,
+        message: &OutboxMessageFull<impl AtomicBatch>,
+    ) -> Result<()> {
+        let mut buffer = Vec::with_capacity(MAX_OUTPUT_SIZE);
+        message
+            .bin_write(&mut buffer)
+            .map_err(|_| OutboxError::OutboxMessageSerializationError)?;
+        rt.write_output(&buffer)?;
+        Ok(())
+    }
+
+    pub fn new(max: u32) -> Self {
+        Self {
+            meta: OutboxQueueMeta { queue_len: 0, max },
+            rollup_outbox_queue: RollupOutboxQueue::new(&ROLLUP_OUTBOX_QUEUE_ROOT, max)
+                .unwrap(),
+        }
+    }
+
+    pub fn queue_len(&self) -> u32 {
+        self.meta.queue_len
+    }
+
+    pub fn incr_queue_len(&mut self) {
+        self.meta.queue_len += 1;
+    }
+
+    pub fn max(&self) -> u32 {
+        self.meta.max
+    }
+
+    /// Flushes the outbox queue in the order of the rollup outbox queue
+    /// then the outbox queue snapshot. The outbox has a maximum capacity
+    /// of 100 messages per level. Messages that cannot be flushed in the
+    /// current level are enqueued into the rollup outbox queue for the
+    /// next flush.
+    pub fn flush(
```

Comment:
Having read the kernel SDK implementation, I'm a bit concerned that we're writing to durable storage too much here, each `rollup_outbox_queue.queue_message` updates the `meta` in durable storage.   

What I suggest is that we leave a comment here (associated with a linear issue) for optimising durable storage writes.

Reply:
Done